tx-compiler/src/compiler.rs

```diff
@@ -0,0 +1,299 @@
+use super::{
+    AccountCode, AccountId, Assembler, AssemblyContext, AssemblyContextType, CodeBlock,
+    CompiledTransaction, Digest, HashMap, Kernel, MidenLib, ModuleAst, Note, NoteScript, Operation,
+    Program, ProgramAst, StdLibrary, TransactionError, TransactionKernel,
+};
+
+// TRANSACTION COMPILER
+// ================================================================================================
+
+pub struct TransactionComplier {
+    assembler: Assembler,
+    account_procedures: HashMap<AccountId, Vec<Digest>>,
+    prologue: CodeBlock,
+    epilogue: CodeBlock,
+    note_setup: CodeBlock,
+    tx_kernel: Kernel,
+}
+
+impl TransactionComplier {
+    // CONSTRUCTOR
+    // --------------------------------------------------------------------------------------------
+    pub fn new() -> TransactionComplier {
+        let (tx_kernel, tx_module) = TransactionKernel::kernel();
+        let assembler = Assembler::default()
+            .with_library(&MidenLib::default())
+            .expect("library is well formed")
+            .with_library(&StdLibrary::default())
+            .expect("library is well formed")
+            .with_kernel_module(tx_module)
+            .expect("kernel is well formed");
+        TransactionComplier {
+            assembler,
+            account_procedures: HashMap::default(),
+            prologue: TransactionKernel::prologue(),
+            epilogue: TransactionKernel::epilogue(),
+            note_setup: TransactionKernel::note_setup(),
+            tx_kernel,
+        }
+    }
+
+    /// Compiles the provided module into [AccountCode] and associates the resulting procedures
+    /// with the specified account ID.
+    pub fn load_account(
+        &mut self,
+        account_id: AccountId,
+        account_code: ModuleAst,
+    ) -> Result<AccountCode, TransactionError> {
+        let account_code = AccountCode::new(account_code, account_id, &mut self.assembler)
```

Comment:
We can run into an issue here if we load the same account twice.  We may want to do this for example when some account code has been updated. This could result in an error if we have some procedures which are the same across the two versions.  As it stands we run into an error where the assembler complains that the procedure already exists.